pydatastructs/miscellaneous_data_structures/segment_tree.py

```diff
@@ -0,0 +1,241 @@
+
+            self._lazy_[_node_] = 0
+
+        if _treeLeft_ >= _left_ and _treeRight_ <= _right_ :
+
+            return self._tree_[_node_]
+
+        elif _treeLeft_ > _right_ or _treeRight_ < _left_ :
+
+            return 0
+
+        else :
+
+            _mid_ = (_treeLeft_ + _treeRight_) // 2
+
+            return \
+                self.__rangeQuery__(2 * _node_, _treeLeft_, _mid_, _left_, _right_) + \
+                self.__rangeQuery__(2 * _node_ + 1, _mid_ + 1, _treeRight_, _left_, _right_)
+
+    """
+        Updates the tree accordingly for the specified range [L, R] for a particular [node] and [newVal].
+    
+        [Internally called by the rangeUpdate(l, r) function]
+    """
+
+    def __rangeUpdate__(self, _node_, _treeLeft_, _treeRight_, _left_, _right_, _newVal_) :
+
+        if self._lazy_[_node_] :
+
+            self._tree_[_node_] += self._lazy_[_node_] * (_treeRight_ - _treeLeft_ + 1)
+
+            if _treeLeft_ != _treeRight_ :
+
+                self._lazy_[2 * _node_] += self._lazy_[_node_]
+                self._lazy_[2 * _node_ + 1] += self._lazy_[_node_]
+
+            self._lazy_[_node_] = 0
+
+        if _treeLeft_ >= _left_ and _treeRight_ <= _right_ :
+
+            self._tree_[_node_] += _newVal_ * (_treeRight_ - _treeLeft_ + 1)
+
+            if _treeLeft_ != _treeRight_ :
+
+                self._lazy_[2 * _node_] += _newVal_
+                self._lazy_[2 * _node_ + 1] += _newVal_
+
+        elif _treeLeft_ > _right_ or _treeRight_ < _left_ :
+
+            return
+
+        else :
+
+            _mid_ = (_treeLeft_ + _treeRight_) // 2
+
+            self.__rangeUpdate__(2 * _node_, _treeLeft_, _mid_, _left_, _right_, _newVal_)
+            self.__rangeUpdate__(2 * _node_ + 1, _mid_ + 1, _treeRight_, _left_, _right_, _newVal_)
+
+            self._tree_[_node_] = self._tree_[2 * _node_] + self._tree_[2 * _node_ + 1]
+
+    """
+        Updates the tree accordingly for the specified position [pos] with the [newVal].
+    
+        [Internally called by the pointUpdate(l, r) function]
+    """
+
+    def __pointUpdate__(self, _pos_, _newVal_) :
+
+        self._tree_[self._treeSize_ // 2 + _pos_] = _newVal_
+
+        _node_ = (self._treeSize_ // 2 + _pos_) // 2
+
+        while _node_ >= 1 :
+
+            self._tree_[_node_] = self._tree_[2 * _node_] + self._tree_[2 * _node_ + 1]
+            _node_ //= 2
+
+    """
+    -- All functions below are the User Functions which invoke the necessary
+       Internal Functions as mentioned earlier. --
+    """
+
+    def rangeUpdate(self, _left_, _right_, _newVal_) :
+
+        self.__rangeUpdate__(1, 0, self._treeSize_ // 2 - 1, _left_, _right_, _newVal_)
+
+    def pointUpdate(self, _pos_, _newVal_) :
+
+        self.__pointUpdate__(_pos_, _newVal_)
+
+    def rangeQuery(self, _left_, _right_) :
+
+        return self.__rangeQuery__(1, 0, self._treeSize_ // 2 - 1, _left_, _right_)
+
+    """
+    Example for using the SegmentTree class (with testcase) :
```

Comment:
By tests I meant in a file like, `test_segment_tree.py`. Would suggest to take a look at `tests` folders in the project. You will get to know about writing tests for an API in Python.

Reply:
Oh...Alright...will add that..